### PR TITLE
fix(web-search): enforce hard content budgets

### LINKED ISSUE #19037: [Feature]: Hard Budget and Predictable Degradation for Built-in Web Search
## Background

Cherry Studio already applies one shared `cutoff_limit` Token budget across all web-search results, with cutoff enabled by default. The remaining risks are legacy settings that migrate to no compression, missing hard protection when fetched content still exceeds the request budget, and unclear degradation when pages are exceptionally large. Semantic compression is a separate P2 proposal.

## Goal

Built-in web search always respects a hard request-content budget and degrades predictably when legacy or exceptional inputs would exceed it.

## Spec

- Preserve the current shared cutoff budget as the default behavior.
- Detect legacy or invalid compression settings that resolve to unbounded content and guide users to a safe value.
- Enforce a final hard content budget before search results enter a model request, including when compression is disabled.
- Degrade oversized results by bounded truncation or omission with visible reasons instead of sending an invalid request.
- Report per-result contribution and the final retained total for diagnosis.
- Track embedding-based semantic compression as a separate optional P2 scope, not a prerequisite for the hard guard.

## Verification

- A migrated legacy profile resolves to no compression -> the user is warned and the final request still cannot exceed the hard budget.
- Several large pages exceed the shared cutoff -> retained content stays within budget and omitted content is identified.
- A page is individually larger than the budget -> the search flow degrades that page without failing the entire Agent request.
- Default settings on a fresh profile keep the existing cutoff behavior.

## Related

#17735

### Source evidence

The following privacy-reviewed source evidence is linked to this requirement.

Source evidence 1

Source evidence 2

Source evidence 3

Source evidence 4

### PR DESCRIPTION
**Pull request by userInner**
> ### Branch strategy
>
> - Active development targets `main`.

### What this PR does

Before this PR:

- The shared cutoff relied on a heuristic token estimate, and `none` compression bypassed content limits entirely.
- A cutoff smaller than the result count could still retain one token per result and exceed the aggregate budget.
- Truncated or omitted content had no structured diagnostics or visible reason in the chat UI.

After this PR:

- Model-bound web search results always pass through aggregate token and UTF-8 byte budgets, including legacy `none` configurations.
- Results degrade in stable order through bounded truncation or omission, with per-result and aggregate diagnostics.
- Truncation reasons reach the model output and appear in the chat result list; settings warn about disabled compression and can restore the default 2,000-token cutoff in one click.
- The explicit `fetchUrlsUnprocessed` knowledge-ingestion path remains raw and unchanged.

Fixes #19037

### Why we need it and why it was done in this way

The final guard must be independent of legacy compression preferences and must not let one unusually large page fail an entire Agent request. The existing equal per-result cutoff allocation is preserved for normal configurations. When the shared limit is smaller than the result count, earlier results receive the available one-token allocations and the remaining results are explicitly marked as omitted.

The following tradeoffs were made:

- Web search post-processing is model-agnostic, so the token contribution uses the conservative maximum of the existing `tokenx` estimate plus the lazily loaded `o200k` and `cl100k` reference encodings. A separate aggregate UTF-8 byte ceiling bounds highly compressible or adversarial content.
- Reference tokenizers are loaded only when web-search post-processing runs, avoiding main-process startup cost.
- Diagnostics are optional additions to the existing output schema, preserving compatibility with stored and provider-native results.

The following alternatives were considered:

- Provider-specific tokenizer plumbing was not added because this service does not know the downstream model dialect and several providers do not expose an exact local tokenizer.
- Embedding-based semantic compression remains separate P2 scope, as requested in the issue.
- Dropping omitted result records entirely was rejected because users and models need a visible reason for missing content.

Links to places where the discussion took place: https://github.com/CherryHQ/cherry-studio/issues/19037

### Breaking changes

None.

### Special notes for your reviewer

- `fetchUrlsUnprocessed` intentionally bypasses the model-bound guard because it feeds knowledge ingestion rather than a model request.
- The conservative budget is covered against numeric runs, `cl100k`/`o200k` differences, tokenizer control text, Unicode surrogate boundaries, multiple oversized results, disabled compression, and sub-result budgets.
- Validation completed: 91 focused tests, `pnpm test:lint`, full type checks, i18n validation for all 12 locales, and `pnpm build`.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Branch: This PR targets `main`
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
Web search now enforces predictable content limits even when compression is disabled, identifies truncated or omitted results, and helps users restore the recommended default cutoff.
```

### CODE REVIEW

**Review by 404-Page-Found — CHANGES_REQUESTED** [review]
Max-effort recall review — 12 findings (see inline comments). Workflow verify phase errored on schema, so findings were manually verified against the fork diff main...pr-19369.

File: src/main/services/webSearch/postProcessing.ts [hunk 1]
```diff
@@ -1,10 +1,80 @@
...
+const MAX_UTF8_BYTES_PER_BUDGET_TOKEN = 8
+const ABSOLUTE_UTF8_BYTE_LIMIT = 400_000
+
+type BudgetTokenizer = {
+  count: (content: string) => number
+}
+
+let budgetTokenizerPromise: Promise<BudgetTokenizer> | undefined
+
+function loadBudgetTokenizer(): Promise<BudgetTokenizer> {
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts:24]: **Cached rejected tokenizer promise never recovers** (CONFIRMED)

`loadBudgetTokenizer` caches via `budgetTokenizerPromise ??= Promise.all(...)` with no reset on rejection. If the first dynamic import of `gpt-tokenizer` fails (missing module, ESM error, transient FS hiccup), the promise stays rejected forever. Every subsequent `postProcessWebSearchResponse` awaits the same rejected promise and throws, turning web search into a hard failure until restart.

Fix: chain `.catch(e => { budgetTokenizerPromise = undefined; throw e })` or reset on rejection, and consider falling back to `estimateTokenCount`-only when the import fails.

File: src/main/services/webSearch/postProcessing.ts [hunk 2]
```diff
@@ -29,30 +99,129 @@ export async function postProcessWebSearchResponse(
...
-  if (runtimeConfig.compression.method === 'cutoff') {
-    return {
-      response: {
-        ...response,
-        results: applyCutoff(response.results, runtimeConfig.compression)
-      }
+  const tokenLimit = Math.max(1, Math.floor(normalizeWebSearchCutoffLimit(runtimeConfig.compression.cutoffLimit)))
+  const byteLimit = Math.min(ABSOLUTE_UTF8_BYTE_LIMIT, tokenLimit * MAX_UTF8_BYTES_PER_BUDGET_TOKEN)
+  const originalBytes = response.results.reduce((total, result) => total + utf8ByteLength(result.content), 0)
+  const tokenizer = await loadBudgetTokenizer()
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts:146]: **Async tokenizer import has no fallback guard** (CONFIRMED)

Previously `applyCutoff` was sync and used only the `sliceByTokens` heuristic. Now `postProcessWebSearchResponse` unconditionally `await loadBudgetTokenizer()` (lazy `o200k_base` + `cl100k_base`) with no `catch`/fallback. If `node_modules` is corrupted or ESM loading fails, the whole post-processing throws and `WebSearchService.searchKeywords` bubbles a terminal tool failure instead of returning degraded truncated results.

Fix: wrap the `await loadBudgetTokenizer()` and keep a heuristic-only fallback path when the import rejects.

File: src/main/services/webSearch/postProcessing.ts [hunk 3]
```diff
@@ -29,30 +99,129 @@ export async function postProcessWebSearchResponse(
...
+  return {
+    response: {
+      ...response,
+      results: processed.results,
+      budget: processed.budget
     }
   }
+}
+
+function distributeSharedLimit(limit: number, resultCount: number): number[] {
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts:172]: **`distributeSharedLimit` drops remainder, wastes budget** (CONFIRMED)

`Math.floor(limit / resultCount)` discards the remainder, so `tokenLimit=5` with 2 results allocates `[2,2]` (total 4) wasting 1 token, and similarly for `byteLimit` up to `resultCount-1` bytes. A result that could have kept one more token is truncated/omitted despite the configured limit allowing it. Distribute the remainder round-robin (first `remainder` results get `+1`).

File: src/main/services/webSearch/postProcessing.ts [hunk 4]
```diff
@@ -1,10 +1,80 @@
...
+  while (end > 0) {
+    try {
+      return decoder.decode(bytes.subarray(0, end))
+    } catch {
+      end -= 1
+    }
+  }
+
+  return ''
+}
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts:100]: **Huge content `originalTokens` undercounted via truncated slice** (CONFIRMED)

`countOriginalTokens` returns `Math.max(estimateTokenCount(content), tokenizer.count(byteBounded))` when `content` exceeds `ABSOLUTE_UTF8_BYTE_LIMIT` (400k), where `byteBounded` is only a `byteAllocation`-sized slice (often 16 bytes for small limits). The heuristic `estimateTokenCount` is exactly what undercounts low-delimiter text (e.g. `'1'.repeat(500_000)`), which is why the reference tokenizer was added — but here the reference count is on the tiny slice (~4 tokens) so the `max` stays ~4k vs true ~125k. `budget.originalTokens` and the `hardLimitExceeded` token check are wrong.

File: src/main/services/webSearch/postProcessing.ts [hunk 5]
```diff
@@ -29,30 +99,129 @@ export async function postProcessWebSearchResponse(
...
-    return {
-      response: {
-        ...response,
-        results: applyCutoff(response.results, runtimeConfig.compression)
-      }
+  const tokenLimit = Math.max(1, Math.floor(normalizeWebSearchCutoffLimit(runtimeConfig.compression.cutoffLimit)))
+  const byteLimit = Math.min(ABSOLUTE_UTF8_BYTE_LIMIT, tokenLimit * MAX_UTF8_BYTES_PER_BUDGET_TOKEN)
+  const originalBytes = response.results.reduce((total, result) => total + utf8ByteLength(result.content), 0)
+  const tokenizer = await loadBudgetTokenizer()
+  const byteAllocations = distributeSharedLimit(byteLimit, response.results.length)
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts]: **Aggregate `originalTokens` uses per-result byte slices, skewing total** (CONFIRMED)

The early `hardLimitExceeded` sum does `sliceByUtf8Bytes(result.content, byteAllocations[index])` then `countOriginalTokens` on that slice per result. With 2 results each 300k bytes and `byteLimit=400k` (`byteAllocations=[200k,200k]`), it counts ~50k tokens per slice vs true ~75k each, sum 100k vs true 150k. The token-limit comparison and reported `budget.originalTokens` are under-reported; the check can miss a token overage that the byte check doesn't catch.

File: src/main/services/webSearch/postProcessing.ts [hunk 6]
```diff
@@ -1,10 +1,80 @@
...
+function sliceByUtf8Bytes(content: string, limit: number): string {
+  const bytes = Buffer.from(content, 'utf8')
+  if (bytes.length <= limit) {
+    return content
+  }
+
+  const decoder = new TextDecoder('utf-8', { fatal: true })
+  let end = limit
+  while (end > 0) {
+    try {
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts:86]: **Bare `catch` in UTF-8 slicer hides unrelated errors** (PLAUSIBLE)

`sliceByUtf8Bytes` does a bare `catch { end -= 1 }` around `decoder.decode(bytes.subarray(0, end))`. If `TextDecoder` throws for a reason other than an incomplete sequence (unsupported encoding, internal error) or `bytes.subarray` throws, the error is silently swallowed and the function returns wrong truncated content instead of surfacing the bug. Narrow the catch to `RangeError`/decode errors.

File: src/main/services/webSearch/postProcessing.ts [hunk 7]
```diff
@@ -29,30 +99,129 @@ export async function postProcessWebSearchResponse(
     return { response }
   }
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts:142]: **Falsy `cutoffLimit` guard dropped — `0` becomes `2000`** (CONFIRMED)

`normalizeWebSearchCutoffLimit(0)` returns `DEFAULT_WEB_SEARCH_CUTOFF_LIMIT` (2000). The old `applyCutoff` had `if (!config.cutoffLimit) return results` (no truncation for `0`/falsy). Now a persisted `cutoffLimit=0` (or `NaN`/`undefined` from corruption) is silently normalized to 2000 and truncates where the caller expected a pass-through. Preserve an explicit `0`/`null` opt-out or document the new semantics.

File: src/main/services/webSearch/postProcessing.ts [hunk 8]
```diff
@@ -29,30 +99,129 @@ export async function postProcessWebSearchResponse(
...
+  const byteAllocations = distributeSharedLimit(byteLimit, results.length)
+  const processed = results.map((result, index) => {
+    const byteBounded = sliceByUtf8Bytes(result.content, byteAllocations[index])
+    const originalTokens = countOriginalTokens(result.content, byteBounded, tokenizer)
+    const content = sliceToBudget(result.content, tokenAllocations[index], byteAllocations[index], tokenizer)
+    const retainedTokens = tokenizer.count(content)
+    const status: WebSearchResultBudget['status'] =
+      content.length === result.content.length ? 'retained' : content.length === 0 ? 'omitted' : 'truncated'
+    const metrics = {
+      originalTokens,
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts:207]: **Truncation ellipsis marker removed from content** (PLAUSIBLE - behavioral change)

Old `applyCutoff` returned `` `${sliced}...` `` when truncated, giving an in-content signal. New `applyContentBudget` returns the raw sliced string with no marker; truncation is only visible via the separate `result.budget` metadata. Any consumer that reads `result.content` without inspecting `result.budget` (cached snapshot, logger, legacy adapter, model prompt that concatenates contents) will treat a truncated excerpt as complete and may hallucinate citations. Consider re-adding a marker or documenting the contract change.

File: src/main/services/webSearch/postProcessing.ts [hunk 9]
```diff
@@ -29,30 +99,129 @@ export async function postProcessWebSearchResponse(
...
+  const originalBytes = response.results.reduce((total, result) => total + utf8ByteLength(result.content), 0)
+  const tokenizer = await loadBudgetTokenizer()
+  const byteAllocations = distributeSharedLimit(byteLimit, response.results.length)
+  const originalTokens = response.results.reduce((total, result, index) => {
+    const byteBounded = sliceByUtf8Bytes(result.content, byteAllocations[index])
+    return total + countOriginalTokens(result.content, byteBounded, tokenizer)
+  }, 0)
+  const hardLimitExceeded = originalTokens > tokenLimit || originalBytes > byteLimit
+
+  if (runtimeConfig.compression.method !== 'cutoff' && !hardLimitExceeded) {
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts:151]: **`compression: 'none'` no longer guarantees raw pass-through** (CONFIRMED - intentional but user-visible)

Old invariant was `method === 'none'` returned results unconditionally. Now `none` still enforces a hard token/byte guard when `hardLimitExceeded` (e.g. a 300k-token scrape with `cutoffLimit=2000`). This is the PR's intent (safety limit), but a user who explicitly selected "None" expecting full page fetch for local summarization will be silently truncated to a 2000-token shared budget with no error unless they notice the new UI warning. Confirm this semantic is intended and surfaced clearly.

File: src/main/services/webSearch/postProcessing.ts [hunk 10]
```diff
@@ -29,30 +99,129 @@ export async function postProcessWebSearchResponse(
...
+    response: {
+      ...response,
+      results: processed.results,
+      budget: processed.budget
     }
   }
+}
+
+function distributeSharedLimit(limit: number, resultCount: number): number[] {
+  if (limit < resultCount) {
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts]: **At-least-one-token per-result guarantee dropped** (CONFIRMED)

`distributeSharedLimit` returns `[1,0,0,0]` when `limit=1` and `resultCount=4`, so trailing results become `''` with `budget.status='omitted'`. Old code used `Math.max(1, floor(limit/count))` guaranteeing at least 1 token per result. The new behavior produces empty cards / zero-content placeholders that the prior contract avoided. If this is intentional, document it; otherwise distribute at least 1 token or omit the result entirely instead of returning `''`.

File: src/main/services/webSearch/postProcessing.ts [hunk 11]
```diff
@@ -1,10 +1,80 @@
...
+  }
+
+  return ''
+}
+
+function countOriginalTokens(content: string, byteBounded: string, tokenizer: BudgetTokenizer): number {
+  return utf8ByteLength(content) <= ABSOLUTE_UTF8_BYTE_LIMIT
+    ? tokenizer.count(content)
+    : Math.max(estimateTokenCount(content), tokenizer.count(byteBounded))
+}
```
Comment (404-Page-Found) [inline src/main/services/webSearch/postProcessing.ts]: **Only high surrogate trimmed, low surrogate remains** (PLAUSIBLE)

`trimDanglingHighSurrogate` checks `0xD800-0xDBFF` and removes a trailing high surrogate, but a lone low surrogate (`\uDC00`, `0xDC00-0xDFFF`) is not removed. An unpaired surrogate in `content` after slicing can break JSON serialization and the renderer's `TextDecoder`. Also guard for the general unpaired case or validate via `encodeURIComponent` check.

File: src/renderer/components/chat/messages/tools/webSearch/MessageWebSearch.tsx [hunk 12]
```diff
@@ -21,6 +22,20 @@ function isRecord(value: unknown): value is Record<string, unknown> {
   return typeof value === 'object' && value !== null && !Array.isArray(value)
 }
 
+function getBudgetStatusKey(budget: WebSearchOutputItem['budget']) {
+  if (!budget || budget.status === 'retained') return undefined
+
+  if (budget.status === 'truncated') {
+    return budget.reason === 'configured_cutoff'
+      ? 'message.websearch.budget.truncated.configured_cutoff'
```
Comment (404-Page-Found) [inline src/renderer/components/chat/messages/tools/webSearch/MessageWebSearch.tsx:30]: **Fallback to `hard_limit` hides missing budget reason** (PLAUSIBLE)

`getBudgetStatusKey` does `budget.reason === 'configured_cutoff' ? cutoffKey : hardLimitKey` for truncated/omitted. If `budget.reason` is `undefined` due to schema evolution, corrupt IPC, or persisted old data, it silently maps to `hard_limit` and shows "Truncated — safety limit" instead of surfacing data corruption. Prefer an explicit fallback or an "unknown" badge when `reason` is missing.

**Review by 404-Page-Found — CHANGES_REQUESTED** [review]
Max-effort recall review — 12 findings posted inline (ranked most severe first).

Top issues: cached rejected tokenizer promise, async import with no fallback, remainder-wasting budget distribution, and under-counting of originalTokens for large content via truncated slices. See each inline comment for file:line, failure scenario, and suggested fix. Workflow verify phase errored on schema so findings were manually verified against `main...pr-19369` (`e3c16d2e`).

File: src/main/services/webSearch/postProcessing.ts (hunk 9 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts:151]: Confirmed as intentional. none disables configured compression, but not the final safety ceiling for model-bound payloads. The settings UI explicitly warns that a final safety limit still applies and offers the default-cutoff action; truncated results also carry structured hard_limit metadata and a visible safety-limit badge.

File: src/main/services/webSearch/postProcessing.ts (hunk 8 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts:207]: The marker removal is intentional. Appending an ellipsis after slicing adds unaccounted tokens and bytes and can exceed the hard limit. Truncation is now represented by structured per-result budget metadata, retained through the tool output schema and rendered as a configured-cutoff or safety-limit badge, so consumers have an explicit machine-readable signal without mutating the bounded content.

File: src/main/services/webSearch/postProcessing.ts (hunk 10 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts]: This behavior is intentional because a shared limit smaller than the result count cannot grant one token to every result without violating the aggregate contract. The zero-allocation entries remain as citation/result metadata with an explicit omitted budget status rather than silently overspending the configured limit.

File: src/renderer/components/chat/messages/tools/webSearch/MessageWebSearch.tsx (hunk 12 above)
Comment (userInner) [reply src/renderer/components/chat/messages/tools/webSearch/MessageWebSearch.tsx:30]: No fallback for malformed data is reachable here: truncated and omitted are a discriminated union whose reason is required and restricted to configured_cutoff or hard_limit in webSearchOutputSchema. Missing or unknown reasons fail safeParse before this renderer path; older results without any budget remain valid and simply show no badge. Therefore the final branch only handles already-validated values.

File: src/main/services/webSearch/postProcessing.ts (hunk 11 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts]: Fixed in signed commit 6111168f. The boundary sanitizer now removes a trailing high surrogate and a trailing low surrogate only when it is not paired with a preceding high surrogate. A regression covers the unpaired-low case.

File: src/main/services/webSearch/postProcessing.ts (hunk 1 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts:24]: Fixed in signed commit 6111168f. The cached promise is cleared on rejection, the current request falls back to the tokenx heuristic instead of failing web search, and the next request retries the reference tokenizer. The new fallback regression test covers both degraded completion and recovery.

File: src/main/services/webSearch/postProcessing.ts (hunk 3 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts:172]: Fixed in signed commit 6111168f. Shared token and byte limits now allocate floor(limit / count) plus one unit to the first remainder entries. The regression for a 5-token budget across two results now retains 3 + 2 tokens, using the full budget.

File: src/main/services/webSearch/postProcessing.ts (hunk 7 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts:142]: This is intentionally unchanged after rechecking the product path. cutoffLimit is normalized at preference/config resolution and in the renderer, and zero/null is not a supported opt-out value. The explicit opt-out is compression method none; even that retains the final hard safety guard. Preserving a corrupt zero as unlimited would reintroduce the unbounded payload this PR fixes.

File: src/main/services/webSearch/postProcessing.ts (hunk 5 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts]: Fixed in signed commit 6111168f. Full-content token counts are computed once for every result before any per-result token or byte allocation is applied, then reused by truncation and aggregate metadata. A two-result 300 KB + 300 KB regression locks this boundary.

File: src/main/services/webSearch/postProcessing.ts (hunk 2 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts:146]: Fixed in signed commit 6111168f. Reference-tokenizer loading now has a heuristic fallback, so a broken or transient dynamic import cannot turn post-processing into a terminal tool failure. The fallback remains byte-bounded, and the promise is reset so a later request can recover.

File: src/main/services/webSearch/postProcessing.ts (hunk 4 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts:100]: Fixed in signed commit 6111168f. originalTokens is now counted from the complete content with the reference encoders, using their generator APIs to avoid building one large token array. A 500,000-character low-delimiter regression verifies that both the per-result and aggregate metrics report the full count.

File: src/main/services/webSearch/postProcessing.ts (hunk 6 above)
Comment (userInner) [reply src/main/services/webSearch/postProcessing.ts:86]: Fixed in signed commit 6111168f. The retry loop now consumes only Node TextDecoder TypeError failures with ERR_ENCODING_INVALID_ENCODED_DATA. Any unrelated exception is rethrown instead of being mistaken for an incomplete UTF-8 boundary.

**Comment by userInner** [comment]
@404-Page-Found Thanks for the detailed review. Signed commit 6111168f addresses the actionable findings: recoverable tokenizer loading with heuristic fallback and retry, exact generator-based full-content token counts, aggregate counts before allocation, remainder-preserving distribution, narrow UTF-8 decode error handling, and trailing unpaired-surrogate protection. I added regressions for failed-import recovery, 500 KB single content, 300 KB + 300 KB aggregate accounting, and 5/2 remainder allocation.

I also replied inline with the verified contract decisions: the hard guard remains active for compression none, a budget smaller than the result count uses explicit omitted entries, the ellipsis is represented by structured budget metadata rather than unbudgeted content, zero is not a supported unlimited preference, and malformed budget reasons are rejected by the discriminated schema before rendering.

Validation on the rebased head: 95 related tests passed, full typecheck passed, and pnpm lint passed. The commit is signed, DCO-signed-off, and GitHub Verified. Could you please re-review the current head when convenient?

**Review by 404-Page-Found — APPROVED** [review]

**Comment by userInner** [comment]
Addressed both hidden cherry-review findings in signed commit f9447ff300. If either reference tokenizer fails to load, the fallback now counts UTF-8 bytes and truncates to the token allocation in bytes; this is intentionally conservative but hard-safe because supported BPE token counts cannot exceed source-byte count, whereas estimateTokenCount could undercount adversarial text. For the normal reference-tokenizer path, truncation now derives concrete prefixes from the heuristic and both encodings and selects the shortest valid prefix, then rechecks the combined counter. The old binary search over BPE prefix counts was removed because that predicate is not monotonic; a byte-prefix fail-closed guard handles any re-encoding edge case. The fallback regression now pins retained tokens/bytes and verifies both reference encodings. All 37 web-search post-processing/service tests pass; ESLint, Biome, and diff checks pass.

File: src/main/services/webSearch/WebSearchService.ts [hunk 13]
```diff
@@ -236,6 +236,18 @@ export class WebSearchService extends BaseService {
     const filteredResponse = filterWebSearchResponseWithBlacklist(mergedResponse, context.runtimeConfig.excludeDomains)
     const postProcessed = await postProcessWebSearchResponse(filteredResponse, context.runtimeConfig)
 
+    if (postProcessed.response.budget) {
```
[the remark this answers is not in the document]
Comment (userInner) [reply src/main/services/webSearch/WebSearchService.ts]: Fixed in c0d1cf930f. Configured cutoff is now treated as the expected path and no longer emits a warning; the warning remains for the fail-safe `hard_limit` path. The service tests now assert both sides of that logging contract.

File: src/shared/ai/builtinTools.ts [hunk 14]
```diff
@@ -410,7 +410,34 @@ export const webSearchOutputItemSchema = z.object({
   id: z.union([z.string(), z.number().int().positive()]),
   title: z.string(),
   url: z.string(),
-  content: z.string()
+  content: z.string(),
+  budget: z
```
[the remark this answers is not in the document]
Comment (userInner) [reply src/shared/ai/builtinTools.ts:423]: Fixed in c0d1cf930f. The four budget metrics and the reason enum now each have one shared Zod schema, while the discriminated branches only add their status/reason shape. This removes the three-way metric duplication without changing the output contract.

**Comment by 404-Page-Found** [comment]
@DeJeune

**Comment by DeJeune** [comment]
> This comment was translated automatically.

The approach isn't very agentic; I'll make some changes.

---
<details>
<summary>Original Content</summary>

做法有点不agentic，我来改一改吧
</details>

**Comment by userInner** [comment]
Synced the latest `main` and resolved the import/rendering conflicts; signed merge commit `0b6f268b2a` is pushed. I saw your note that the approach could be made more agentic, so I am leaving the implementation open for your follow-up changes.

**Comment by userInner** [comment]
Fixed the failed `basic-checks` job in signed commit `3986ce1846` (GitHub Verified). The check was failing because `pnpm i18n:sync` added seven new web-search budget/compression keys to `tr-tr.json` but the generated locale changes were not committed. Committed the synchronized Turkish catalog; no runtime logic changed. Please re-run CI on the new head.

**Comment by userInner** [comment]
> This comment was translated automatically.

The CI's i18n check found that the newly added Turkish entries still contain [to be translated] markers. I have added 7 official Turkish translations in signed commit 0e9fe6a986, and passed i18n:check, i18n:unused:check, and diff checks locally. Please re-run the CI.

---
<details>
<summary>Original Content</summary>

CI 的 i18n 检查发现土耳其语新增条目仍包含 [to be translated] 标记。我已在签名提交 0e9fe6a986 中补齐 7 条正式土耳其语翻译，并在本地通过 i18n:check、i18n:unused:check 与 diff 检查。请重新运行 CI。
</details>
